### PR TITLE
Add db benchmark script

### LINKED ISSUE #1870: Add a script for running full db-benchmark suite
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)
I've been working on db-benchmark and think it would be valuable to automate the full process so others could run the benchmarks as well

**Describe the solution you'd like**
A new shell script in `benchmarks/db-benchmark` that runs the full db-benchmark process

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by matthewmturner**
# Which issue does this PR close?

Closes #1870 

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by matthewmturner** [comment]
meh, i think im close but im having some issues due to docker memory swapping which isnt allowed by some db-benchmark helpers.  will come back to this.

**Comment by matthewmturner** [comment]
Lol. I unknowingly created a shell script that had a name conflict with a db-benchmark script which caused issue.

File: benchmarks/db-benchmark/db-benchmark.dockerfile [hunk 1]
```diff
@@ -0,0 +1,54 @@
+FROM ubuntu
+ARG DEBIAN_FRONTEND=noninteractive
+
+RUN apt-get update && \
+    apt-get install -y git build-essential
+
+# Install R, curl, and python deps
+RUN apt-get update && apt-get -y install --no-install-recommends --no-install-suggests \
```
Comment (houqp) [inline benchmarks/db-benchmark/db-benchmark.dockerfile:8]: the apt-get update seems to be redundant? looks like we can merge these two apt runs to reduce build time.

File: benchmarks/db-benchmark/db-benchmark.dockerfile [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+RUN git clone https://github.com/h2oai/db-benchmark \
+    && Rscript db-benchmark/_data/groupby-datagen.R 1e7 1e2 0 0 \
+    && Rscript db-benchmark/_data/join-datagen.R 1e7 0 0 0
+
+# Copy local arrow-datafusion
+COPY . arrow-datafusion
+
+# Clone datafusion-python and build python library
+# Not sure if the wheel will be the same on all computers
+RUN git clone https://github.com/datafusion-contrib/datafusion-python \
```
Comment (houqp) [inline benchmarks/db-benchmark/db-benchmark.dockerfile:35]: would be good to clone a particular tag/commit to make this more reproducible.

File: benchmarks/db-benchmark/db-benchmark.dockerfile (hunk 1 above)
Comment (matthewmturner) [reply benchmarks/db-benchmark/db-benchmark.dockerfile:8]: Good catch. Thx.

**Comment by matthewmturner** [comment]
I have this working now.

Would at least one other person be able to give it a try.  I do the following from root directory of `arrow-datafusion`

```
$ docker build -t db-benchmark -f benchmarks/db-benchmark/db-benchmark.dockerfile .
# I used privileged as its supposed to improve with CPU intensive scripts
$ docker run --privileged db-benchmark
```

So far im not seeing as good of results in docker as i was natively but i think thats expected, even using `--privileged`.  I havent had to performance tune docker too much in the past though so if anyone else has other ideas im open to it.

Right now, this only works with official datafusion release (version 7).  However, I added option to pull in local datafusion so that its easier to benchmark local changes.  Even if the docker performance isnt optimal it should at least give a baseline for making local changes and seeing if there are improvements - which i think is the intent of this script.

let me know if any questions or if you see any areas for improvement.

**Comment by houqp** [comment]
@matthewmturner you are on a mac machine right?

**Comment by matthewmturner** [comment]
> @matthewmturner you are on a mac machine right?

Yes, M1 Mac

**Comment by houqp** [comment]
that's expected then. the overhead should go away with when it's executed on a linux box.

**Comment by bobtins** [comment]
Is this specific to M1 Mac? I am running Ubuntu 20.04 on AMD64 and I get this error on docker build:
```
⚠️  Warning: No compatible platform tag found, using the linux tag instead. You won't be able to upload those wheels to PyPI.
📦 Built wheel for abi3 Python ≥ 3.6 to /datafusion-python/target/wheels/datafusion-0.4.0-cp36-abi3-linux_x86_64.whl
WARNING: Requirement 'target/wheels/datafusion-0.4.0-cp36-abi3-linux_aarch64.whl' looks like a filename, but the file does not exist
ERROR: datafusion-0.4.0-cp36-abi3-linux_aarch64.whl is not a supported wheel on this platform.
The command '/bin/sh -c cd datafusion-python     && maturin build --release     && python3 -m pip install target/wheels/datafusion-0.4.0-cp36-abi3-linux_aarch64.whl     && cd ..' returned a non-zero code: 1
```

**Comment by matthewmturner** [comment]
@bobtins thanks for trying it out! You'll just need to update the wheel file name in the docker file. Looks like the below is what you'll need. 

/datafusion-python/target/wheels/datafusion-0.4.0-cp36-abi3-linux_x86_64.whl

**Comment by bobtins** [comment]
Yeah, that got it to work.
I found [this](https://docs.docker.com/engine/reference/builder/#automatic-platform-args-in-the-global-scope) for populating env variables automatically, but it requires BuildKit, which means I have to do `export DOCKER_BUILDKIT=1` before running the build step.

**Comment by matthewmturner** [comment]
@houqp If no other feedback I think this is good to merge for now.

@bobtins im a bit constrained on time right now to work on getting the architectures / wheels integrated properly into the script.  given youve shown interest in the benchmarking area would you be willing to handle that as a follow on PR?

**Comment by yjshen** [comment]
> /datafusion-python/target/wheels/datafusion-0.4.0-cp36-abi3-linux_x86_64.whl

**Comment by bobtins** [comment]
Almost got it...will post a diff as soon as it works.

File: benchmarks/db-benchmark/README.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+-->
+
+# Run db-benchmark
+
+## Directions
+
+Run the following from root `arrow-datafusion` directory
+
+```bash
+$ docker build -t db-benchmark -f benchmarks/db-benchmark/db-benchmark.dockerfile .
```
Comment (bobtins) [inline benchmarks/db-benchmark/README.md:27]: docker **buildx** build ...

**Review by bobtins — CHANGES_REQUESTED** [review]
Got it working with changes suggested.
Looking forward to playing around with this and maybe adding some more features.

File: benchmarks/db-benchmark/db-benchmark.dockerfile [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    && python3 -m pip install -r requirements.txt \
+    && cd ..
+
+# Copy local arrow-datafusion
+COPY . arrow-datafusion
+
+# 1. datafusion-python that builds from datafusion version referenced datafusion-python
+RUN cd datafusion-python \
+    && maturin build --release \
+    && python3 -m pip install target/wheels/datafusion-0.4.0-cp36-abi3-linux_aarch64.whl \
```
Comment (bobtins) [inline benchmarks/db-benchmark/db-benchmark.dockerfile:70]:
```suggestion
    && case "${TARGETPLATFORM}" in \
         */amd64)  CPUARCH=x86_64  ;; \
         */arm64)  CPUARCH=aarch64  ;; \
         *) exit 1 ;; \
    esac \
    && python3 -m pip install target/wheels/datafusion-0.4.0-cp36-abi3-linux_${CPUARCH}.whl \
    && case "${TARGETPLATFORM}" in \
         */amd64)  CPUARCH=x86_64  ;; \
         */arm64)  CPUARCH=aarch64  ;; \
         *) exit 1 ;; \
    esac \
    && python3 -m pip install target/wheels/datafusion-0.4.0-cp36-abi3-linux_${CPUARCH}.whl \
```

Comment (bobtins) [reply benchmarks/db-benchmark/db-benchmark.dockerfile:70]: that's weird, didn't mean to put it in twice; haven't used this github feature before

Comment (bobtins) [reply benchmarks/db-benchmark/db-benchmark.dockerfile:70]: OK, ignore those suggestions...I just did a diff and attached it.

[diff.txt](https://github.com/apache/arrow-datafusion/files/8257176/diff.txt)

**Comment by matthewmturner** [comment]
@bobtins thank you for the help! I am going to review and integrate tomorrow morning.

File: benchmarks/db-benchmark/README.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+-->
+
+# Run db-benchmark
+
+## Directions
+
+Run the following from root `arrow-datafusion` directory
+
+```bash
+$ docker buildx -t db-benchmark -f benchmarks/db-benchmark/db-benchmark.dockerfile .
```
Comment (yjshen) [inline benchmarks/db-benchmark/README.md:27]:
```suggestion
$ docker buildx build -t db-benchmark -f benchmarks/db-benchmark/db-benchmark.dockerfile .
```

Comment (matthewmturner) [reply benchmarks/db-benchmark/README.md:27]: Ah sry, did that local but forgot to add there

**Review by yjshen — APPROVED** [review]
LGTM. Thanks @matthewmturner !

**Pull request merged**